radio/src/crsf.h

```diff
@@ -0,0 +1,143 @@
+#pragma once
+
+#include "opentx.h"
+#include "trainer.h"
+#include "crc.h"
+
+#include <algorithm>
+#include <limits>
+
+void processCrsfInput();
+void crsfTrainerPauseCheck();
+
+#ifdef __GNUC__
+# pragma GCC diagnostic push
+# pragma GCC diagnostic error "-Wswitch" // unfortunately the project uses -Wnoswitch
+#endif
+
+namespace CRSF {
+    static constexpr uint32_t baudrate{420000};
+    void crsfTrainerOnIdle();
+    struct Servo {
+        using Crsf = Trainer::Protocol::Crsf;
+        using MesgType = Crsf::MesgType;
+        
+        static constexpr uint8_t mPauseCount{2}; // 2ms
+        
+        static constexpr uint8_t CRSFChannels{16};
+        
+        enum class State : uint8_t {Undefined, GotFCAddress, GotLength, Channels, Data, AwaitCRC, AwaitCRCAndDecode};
+        
+        static inline int16_t convertCrsfToPuls(uint16_t const value) {
+            const int16_t centered = value - Crsf::CenterValue;
+            return Trainer::clamp((centered * 5) / 8);
+        }
+        
+        inline void tick1ms() {
+            if (mPauseCounter > 0) {
+                --mPauseCounter;
+            }
+            else {
+                mState = State::Undefined;
+            }
+        }
+        
+        inline void process(const uint8_t b, const std::function<void()> f) {
+            mPauseCounter = mPauseCount;
+            ++mBytesCounter;
+            switch(mState) { // enum-switch -> no default (intentional)
+            case State::Undefined:
+                csum.reset();
+                if (b == Crsf::FcAddress) {
```

Comment:
Please note that although ExpressLRS receivers currently output this as their first byte, it is incorrect CRSF and will be changed at some point. Correct CRSF starts with 0xC8 **always** (`UART_SYNC` in the Crossfire units). Betaflight, iNav, ExpressLRS, EdgeTX and many other systems don't check the first byte or check it against a variety of values to validate the frame.

Also note there's a perfectly good CRSF frame parser already in pulses/Crossfire.cpp so it would be nice to not duplicate the code, and with different behavior between them.

I think ideally ExpressLRS would love to be able to send CHANNELSPACKED/SUBSETPACKED on its telemetry connection as well for use as a trainer or to forward headtracking data directly from the TX module's backpack into EdgeTX.

Reply:
That last part sounds interesting